File: sim/request/styles.go
package request

// The fight styles.
//
// A style is a page preset: a name a player picks, and the encounter
// fields it stands for. The expansion lives here, in Go, because the
// page applies it too and a second copy of the table in TypeScript
// would drift the first time a preset was retuned - the page would
// offer "Heavy movement" and the sim would run last month's numbers.
// So this map is the source, `go run ./internal/genstyles` renders it
// to styles.json, and the web lane's test compares its own table
// against that file.
//
// A style sets only the fields it owns. Fight length, variation, target
// level, armor and type are the player's and survive a style change,
// which is what makes the style control a preset rather than a reset.

import (
	"encoding/json"
	"slices"

	"github.com/jhunthrop/foreversixty/sim/api"
)

// style is one preset's expansion, as a delta over the base encounter.
type style struct {
	Targets         int               `json:"targets"`
	ExecuteRatio    float64           `json:"execute_ratio"`
	Movement        *api.Movement     `json:"movement,omitempty"`
	TargetsOverTime []api.TargetCount `json:"targets_over_time,omitempty"`
	Dummy           bool              `json:"dummy,omitempty"`
}

// StyleIDs is the vocabulary, in the order the page lists it.
var StyleIDs = []string{
	"patchwerk",
	"execute",
	"light-movement",
	"heavy-movement",
	"cleave-2",
	"cleave-3",
	"cleave-5",
	"dungeon",
	"dummy",
}

// The movement windows the two movement styles use. Five seconds is
// one global cooldown plus a step; the interval is what makes light
// and heavy different.
const (
	movementWindowSec   = 5
	lightMovementEvery  = 45
	heavyMovementEvery  = 20
	patchwerkExecute    = 0.25
	executeHeavyExecute = 0.35
)

// styles is the contract's section 1.6 table.
var styles = map[string]style{
	"patchwerk": {Targets: 1, ExecuteRatio: patchwerkExecute},
	"execute":   {Targets: 1, ExecuteRatio: executeHeavyExecute},
	"light-movement": {Targets: 1, ExecuteRatio: patchwerkExecute,
		Movement: &api.Movement{IntervalSec: lightMovementEvery, DurationSec: movementWindowSec, Kind: api.MovementAway}},
	"heavy-movement": {Targets: 1, ExecuteRatio: patchwerkExecute,
		Movement: &api.Movement{IntervalSec: heavyMovementEvery, DurationSec: movementWindowSec, Kind: api.MovementAway}},
	"cleave-2": {Targets: 2, ExecuteRatio: patchwerkExecute},
	"cleave-3": {Targets: 3, ExecuteRatio: patchwerkExecute},
	"cleave-5": {Targets: 5, ExecuteRatio: patchwerkExecute},
	// A boss, then packs of three and five, then back down: the shape
	// of a dungeon pull. No execute window, because a pack dies from
	// full health to nothing rather than sliding down a health bar.
	"dungeon": {Targets: 1, ExecuteRatio: 0, TargetsOverTime: []api.TargetCount{
		{AtSec: 0, Count: 1},
		{AtSec: 40, Count: 3},
		{AtSec: 80, Count: 5},
		{AtSec: 130, Count: 3},
		{AtSec: 160, Count: 1},
	}},
	// A training dummy has no debuffs, no execute and no armor
	// reduction; the engine's target_dummy flag is what turns all
	// three off at once.
	"dummy": {Targets: 1, ExecuteRatio: 0, Dummy: true},
}

// ExpandStyle applies a style to an encounter, keeping every field the
// style does not own. The label rides along in Style so a saved sim can
// say which preset produced it.
func ExpandStyle(id string, base api.EncounterSpec) (api.EncounterSpec, bool) {
	s, ok := styles[id]
	if !ok {
		return base, false
	}
	out := base
	out.Style = id
	out.Targets = s.Targets
	out.ExecuteRatio = s.ExecuteRatio
	out.Dummy = s.Dummy
	out.Movement = nil
	if s.Movement != nil {
		// A copy: the map's pointer is shared by every caller, and a
		// caller that edited the encounter it got back would retune the
		// preset for the whole process.
		m := *s.Movement
		out.Movement = &m
	}
	out.TargetsOverTime = slices.Clone(s.TargetsOverTime)
	return out, true
}

// StylesJSON renders the table as the web lane's fixture: an array in
// StyleIDs order, each entry the style id and its expansion.
func StylesJSON() ([]byte, error) {
	type row struct {
		ID string `json:"id"`
		style
	}
	out := make([]row, 0, len(StyleIDs))
	for _, id := range StyleIDs {
		out = append(out, row{ID: id, style: styles[id]})
	}
	b, err := json.MarshalIndent(out, "", "  ")
	if err != nil {
		return nil, err
	}
	return append(b, '\n'), nil
}

File: web/src/lib/sim/styles.ts
// web/src/lib/sim/styles.ts
// The fight styles, contract 1.6. A style is a page preset: it expands to encounter fields
// and leaves its own id behind as `encounter.style`, which is a label the engine ignores.
// The table is duplicated as src/fixtures/sim/styles.json because the sim module lane owns
// a Go copy of it; styles.test.ts asserts this file and that one agree, so the fixture --
// not either implementation -- is the artefact the two lanes share.
//
// Nothing here holds a word: labels and notes are simCopy.styleLabel and simCopy.styleNote,
// keyed by these ids, so a copy change is one diff in copy.ts.
import type { EncounterSpec, Movement, TargetCount } from './types';

export type FightStyleId =
  | 'patchwerk'
  | 'execute'
  | 'light-movement'
  | 'heavy-movement'
  | 'cleave-2'
  | 'cleave-3'
  | 'cleave-5'
  | 'dungeon'
  | 'dummy';

export interface FightStyle {
  id: FightStyleId;
  targets: number;
  execute_ratio: number;
  movement: Movement | null;
  targets_over_time: TargetCount[] | null;
  dummy: boolean;
}

/** The style a fresh settings state opens on. */
export const DEFAULT_STYLE_ID: FightStyleId = 'patchwerk';

const PLAIN_EXECUTE = 0.25;

/** A 5 s window out of melee, at the interval the style names. */
function away(intervalSec: number): Movement {
  return { interval_sec: intervalSec, duration_sec: 5, kind: 'away' };
}

function cleave(id: FightStyleId, targets: number): FightStyle {
  return {
    id,
    targets,
    execute_ratio: PLAIN_EXECUTE,
    movement: null,
    targets_over_time: null,
    dummy: false,
  };
}

export const FIGHT_STYLES: readonly FightStyle[] = [
  {
    id: 'patchwerk',
    targets: 1,
    execute_ratio: PLAIN_EXECUTE,
    movement: null,
    targets_over_time: null,
    dummy: false,
  },
  {
    id: 'execute',
    targets: 1,
    execute_ratio: 0.35,
    movement: null,
    targets_over_time: null,
    dummy: false,
  },
  {
    id: 'light-movement',
    targets: 1,
    execute_ratio: PLAIN_EXECUTE,
    movement: away(45),
    targets_over_time: null,
    dummy: false,
  },
  {
    id: 'heavy-movement',
    targets: 1,
    execute_ratio: PLAIN_EXECUTE,
    movement: away(20),
    targets_over_time: null,
    dummy: false,
  },
  cleave('cleave-2', 2),
  cleave('cleave-3', 3),
  cleave('cleave-5', 5),
  {
    id: 'dungeon',
    targets: 1,
    execute_ratio: 0,
    movement: null,
    targets_over_time: [
      { at_sec: 0, count: 1 },
      { at_sec: 40, count: 3 },
      { at_sec: 80, count: 5 },
      { at_sec: 130, count: 3 },
      { at_sec: 160, count: 1 },
    ],
    dummy: false,
  },
  {
    id: 'dummy',
    targets: 1,
    execute_ratio: 0,
    movement: null,
    targets_over_time: null,
    dummy: true,
  },
];

/** The style with this id, or null. An unknown id is never silently treated as Patchwerk. */
export function fightStyle(id: string): FightStyle | null {
  return FIGHT_STYLES.find((style) => style.id === id) ?? null;
}

/**
 * What TARGETS should say about this encounter (tank MAJOR, review.md:325-327): a
 * timeline style's `targets` field is only the ramp's opening count (the dungeon pull
 * settles on 1, not the 5 it actually reaches), so a control reading that field alone
 * understates the run. This file holds no words -- the caller turns these numbers into a
 * sentence via copy.ts.
 *
 * `attached` (fix round 1, reviewer Important): `settings.ts`'s `detached()` never clears
 * `targets_over_time` when a style-owned field like the dummy checkbox is toggled by hand
 * -- clearing it would silently drop the ramp from the run the moment a player ticks a
 * checkbox, which is exactly the kind of dishonesty this whole fix round exists to remove.
 * So a ramp can outlive its style: `timeline` stays keyed on the ramp alone (the control
 * must never show a number the run will not use), and `attached` is a second, independent
 * flag the caller uses only to pick which sentence is still true, not to change whether
 * the control is read-only.
 */
export function targetsSummary(
  encounter: EncounterSpec,
): { timeline: true; first: number; max: number; attached: boolean } | { timeline: false; count: number } {
  const timeline = encounter.targets_over_time;
  if (timeline !== undefined && timeline.length > 0) {
    return {
      timeline: true,
      first: timeline[0].count,
      max: Math.max(...timeline.map((step) => step.count)),
      attached: (encounter.style ?? '') !== '',
    };
  }
  return { timeline: false, count: encounter.targets };
}

/**
 * The style's fields written over an encounter. Every field a style owns is written on
 * every call, including the nulls and the false: switching from Heavy movement to
 * Patchwerk has to clear the movement block, and a partial write would leave the previous
 * style's fight running under the new style's name.
 *
 * Fight length and duration variation are the player's, not the style's, so they survive.
 */
export function applyFightStyle(encounter: EncounterSpec, id: FightStyleId): EncounterSpec {
  const style = fightStyle(id);
  if (style === null) return encounter;
  return {
    ...encounter,
    style: style.id,
    targets: style.targets,
    execute_ratio: style.execute_ratio,
    // `EncounterSpec.movement`/`targets_over_time` are `omitempty` on the Go side: the
    // wire shape the contract specifies is the key absent, not `null`. `undefined` is the
    // only value that reproduces that -- `JSON.stringify` drops an `undefined` field but
    // still emits `"movement":null` for an explicit null.
    movement: style.movement === null ? undefined : { ...style.movement },
    targets_over_time:
      style.targets_over_time === null ? undefined : style.targets_over_time.map((step) => ({ ...step })),
    dummy: style.dummy,
  };
}
